execution_engine/src/storage/trie_store/lmdb.rs

```diff
@@ -161,3 +172,150 @@ impl<K, V> Store<Digest, Trie<K, V>> for LmdbTrieStore {
 }
 
 impl<K, V> TrieStore<K, V> for LmdbTrieStore {}
+
+pub(crate) type Cache = Arc<Mutex<HashMap<Digest, (bool, Trie<Key, StoredValue>)>>>;
+/// In-memory cached trie store, backed by rocksdb.
+#[derive(Clone)]
+pub struct ScratchCache {
+    pub(crate) cache: Cache,
+    pub(crate) store: Arc<LmdbTrieStore>,
+    pub(crate) env: Arc<LmdbEnvironment>,
+}
+
+/// Cached version of the trie store.
+#[derive(Clone)]
+pub struct ScratchTrieStore {
+    pub(crate) inner: ScratchCache,
+}
+
+impl ScratchTrieStore {
+    /// Creates a new ScratchTrieStore.
+    pub fn new(store: Arc<LmdbTrieStore>, env: Arc<LmdbEnvironment>) -> Self {
+        Self {
+            inner: ScratchCache {
+                store,
+                env,
+                cache: Default::default(),
+            },
+        }
+    }
+
+    /// Writes all dirty (modified) cached tries to underlying db.
```

Comment:
I think this comment is a bit misleading - it certainly fooled me for a while! :)  Rather than writing all dirty tries, it only writes those dirty tries which fall under `new_state_root`.

It's maybe also worth mentioning that all children under `new_state_root` are either read or written, and if reading uncovers a missing child, the method returns an error.

Reply:
We may have to rewrite this to obey the invariant that the trie must never have a node which has missing descendants...

Reply:
Renamed and rewrote the doc-comment here, but we will revisit the implementation.

Reply:
Implementation revisited and resolved in [4b78471](https://github.com/casper-network/casper-node/pull/2877/commits/4b78471c0fbac1bcaf8999792dcda1d0ded85d70)